src/shared/filesystemUtilities.ts

```diff
@@ -198,7 +196,44 @@ export function getFileDistance(fileA: string, fileB: string): number {
  * @param suffix  Filename suffix, typically an extension (".txt"), may be empty
  * @param max  Stop searching if all permutations up to this number exist
  */
-export function getNonexistentFilename(dir: string, name: string, suffix: string, max: number = 99): string {
+export async function getNonexistentFilename(
+    dir: string,
+    name: string,
+    suffix: string,
+    max: number = 99
+): Promise<string> {
+    if (!name) {
+        throw new Error(`name is empty`)
+    }
+    if (!(await fsCommon.existsDir(dir))) {
+        throw new Error(`directory does not exist: ${dir}`)
+    }
+    for (let i = 0; true; i++) {
+        const filename =
+            i === 0 ? `${name}${suffix}` : `${name}-${i < max ? i : crypto.randomBytes(4).toString('hex')}${suffix}`
+        const fullpath = path.join(dir, filename)
+        if (!(await fsCommon.existsFile(fullpath)) || i >= max + 99) {
+            return filename
+        }
+    }
+}
+
+/**
+ * @deprecated this is a synchronous duplicate of {@link getNonexistentFilename}. We are only keeping it
+ * since some code needs to do this process synchronously and the platform agnostic file system is async.
+ *
+ * Returns `name.suffix` if it does not already exist in directory `dir`, else appends
+ * a number ("foo-1.txt", "foo-2.txt", etc.).
+ *
+ * To avoid excessive filesystem activity, if all filenames up to `max` exist,
+ * the function instead appends a random string.
+ *
+ * @param dir  Path to a directory
+ * @param name  Filename without extension
+ * @param suffix  Filename suffix, typically an extension (".txt"), may be empty
+ * @param max  Stop searching if all permutations up to this number exist
+ */
+export function getNonexistentFilenameSync(dir: string, name: string, suffix: string, max: number = 99): string {
```

Comment:
To avoid code duplication, we could pass in `existsDir` and `existsDirFile` as parameters. 

To unblock this PR, I can do that as a followup

Reply:
I'll merge this for now. I'm not sure `existsDirFile` will work with the current logic since it checks different files in a loop.

Reply:
Fixed in https://github.com/aws/aws-toolkit-vscode/pull/4576